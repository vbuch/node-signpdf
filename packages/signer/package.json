{
    "name": "@signpdf/signer",
    "version": "3.2.3",
    "description": "Signer base implementations.",
    "repository": {
        "type": "git",
        "url": "https://github.com/vbuch/node-signpdf"
    },
    "license": "MIT",
    "keywords": [
        "sign",
        "pdf",
        "node",
        "nodejs",
        "esign",
        "adobe",
        "ppklite",
        "sign detached",
        "pkcs7",
        "pkcs#7",
        "pades",
        "digital signature"
    ],
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "files": [
        "dist",
        "LICENSE",
        "README.md"
    ],
    "engines": {
        "node": ">=12",
        "yarn": ">=1.22.18"
    },
    "scripts": {
        "test": "jest",
        "build": "rm -rf ./dist/* & babel ./src -d ./dist --ignore \"**/*.test.js\" & tsc",
        "lint": "eslint -c .eslintrc --ignore-path ../../.eslintignore ./"
    },
    "dependencies": {
        "@signpdf/utils": "^3.2.3"
    },
    "peerDependencies": {
        "asn1js": "^3.0.5",
        "pkijs": "^3.0.15"
    },
    "devDependencies": {
        "@babel/cli": "^7.0.0",
        "@babel/core": "^7.4.0",
        "@babel/eslint-parser": "^7.16.3",
        "@babel/node": "^7.0.0",
        "@babel/plugin-syntax-object-rest-spread": "^7.0.0",
        "@babel/preset-env": "^7.4.2",
        "@signpdf/eslint-config": "^3.2.3",
        "@signpdf/internal-utils": "^3.2.3",
        "@types/node": ">=12.0.0",
        "assertion-error": "^1.1.0",
        "babel-jest": "^27.3.1",
        "babel-plugin-module-resolver": "^3.1.1",
        "coveralls": "^3.0.2",
        "eslint": "^8.50.0",
        "eslint-config-airbnb-base": "^15.0.0",
        "eslint-import-resolver-babel-module": "^5.3.1",
        "eslint-plugin-import": "^2.25.3",
        "eslint-plugin-jest": "^25.2.4",
        "husky": "^7.0.4",
        "jest": "^27.3.1",
        "asn1js": "^3.0.5",
        "pkijs": "^3.0.15",
        "pdfkit": "~0.10.0",
        "typescript": "^5.2.2"
    },
    "gitHead": "5d5ec00c21e072613acb9776c7c6ac7697314955"
}
