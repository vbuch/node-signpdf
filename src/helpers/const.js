export const DEFAULT_SIGNATURE_LENGTH = 8192;
export const DEFAULT_BYTE_RANGE_PLACEHOLDER = '**********';
export const SUBFILTER_ADOBE_PKCS7_DETACHED = 'adbe.pkcs7.detached';
export const SUBFILTER_ADOBE_PKCS7_SHA1 = 'adbe.pkcs7.sha1';
export const SUBFILTER_ADOBE_X509_SHA1 = 'adbe.x509.rsa.sha1';
export const SUBFILTER_ETSI_CADES_DETACHED = 'ETSI.CAdES.detached';
