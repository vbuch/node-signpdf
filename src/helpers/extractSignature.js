import SignPdfError from '../SignPdfError';

const getSubstringIndex = (str, substring, n) => {
    let times = 0; let
        index = null;

    while (times < n && index !== -1) {
        index = str.indexOf(substring, index + 1);
        times += 1;
    }

    return index;
};
/**
 * Basic implementation of signature extraction.
 *
 * Really basic. Would work in the simplest of cases where there is only one signature
 * in a document and ByteRange is only used once in it.
 *
 * @param {Buffer} pdf
 * @returns {Object} {ByteRange: Number[], signature: Buffer, signedData: Buffer}
 */
const extractSignature = (pdf, signatureCount = 1) => {
    if (!(pdf instanceof Buffer)) {
        throw new SignPdfError(
            'PDF expected as Buffer.',
            SignPdfError.TYPE_INPUT,
        );
    }

    // const byteRangePos = pdf.indexOf('/ByteRange [');
    const byteRangePos = getSubstringIndex(pdf, '/ByteRange [', signatureCount);
    if (byteRangePos === -1) {
        throw new SignPdfError(
            'Failed to locate ByteRange.',
            SignPdfError.TYPE_PARSE,
        );
    }

    const byteRangeEnd = pdf.indexOf(']', byteRangePos);
    if (byteRangeEnd === -1) {
        throw new SignPdfError(
            'Failed to locate the end of the ByteRange.',
            SignPdfError.TYPE_PARSE,
        );
    }

    const byteRange = pdf.slice(byteRangePos, byteRangeEnd + 1).toString();
    const matches = (/\/ByteRange \[(\d+) +(\d+) +(\d+) +(\d+) *\]/).exec(byteRange);
    if (matches === null) {
        throw new SignPdfError(
            'Failed to parse the ByteRange.',
            SignPdfError.TYPE_PARSE,
        );
    }

    const ByteRange = matches.slice(1).map(Number);
    const signedData = Buffer.concat([
        pdf.slice(ByteRange[0], ByteRange[0] + ByteRange[1]),
        pdf.slice(ByteRange[2], ByteRange[2] + ByteRange[3]),
    ]);

    const signatureHex = pdf.slice(ByteRange[0] + ByteRange[1] + 1, ByteRange[2])
        .toString('binary')
        .replace(/(?:00|>)+$/, '');

    const signature = Buffer.from(signatureHex, 'hex').toString('binary');

    return {
        ByteRange: matches.slice(1, 5).map(Number),
        signature,
        signedData,
    };
};

export default extractSignature;
