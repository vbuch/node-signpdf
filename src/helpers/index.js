export {default as extractSignature} from './extractSignature';
export {default as pdfkitAddPlaceholder} from './pdfkitAddPlaceholder';
export {default as plainAddPlaceholder} from './plainAddPlaceholder';
export {default as removeTrailingNewLine} from './removeTrailingNewLine';
export {default as findByteRange} from './findByteRange';

'This string is added so that jest collects coverage for this file'; // eslint-disable-line
