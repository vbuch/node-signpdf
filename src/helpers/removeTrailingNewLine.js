import SignPdfError from '../SignPdfError';

/**
 * Removes a trailing new line if there is such.
 *
 * Also makes sure the file ends with an EOF line as per spec.
 * @param {Buffer} pdf
 * @returns {Buffer}
 */
const removeTrailingNewLine = (pdf) => {
    if (!(pdf instanceof Buffer)) {
        throw new SignPdfError(
            'PDF expected as Buffer.',
            SignPdfError.TYPE_INPUT,
        );
    }

    const lastChar = pdf.slice(pdf.length - 1).toString();
    let output = pdf;
    if (lastChar === '\n') {
        // remove the trailing new line
        output = pdf.slice(0, pdf.length - 1);
    }

    const lastLine = output.slice(output.length - 6).toString();
    if (lastLine !== '\n%%EOF') {
        output = Buffer.concat([
            output,
            Buffer.from('\n%%EOF'),
        ]);
    }

    return output;
};

export default removeTrailingNewLine;
