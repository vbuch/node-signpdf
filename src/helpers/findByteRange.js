import SignPdfError from '../SignPdfError';
import {DEFAULT_BYTE_RANGE_PLACEHOLDER} from './const';

/**
 * Finds ByteRange information within a given PDF Buffer if one exists
 *
 * @param {Buffer} pdf
 * @returns {Object} {byteRangePlaceholder: String, byteRangeStrings: String[], byteRange: String[]}
 */
const findByteRange = (pdf) => {
    if (!(pdf instanceof Buffer)) {
        throw new SignPdfError(
            'PDF expected as Buffer.',
            SignPdfError.TYPE_INPUT,
        );
    }

    const byteRangeStrings = pdf.toString().match(/\/ByteRange\s*\[{1}\s*(?:(?:\d*|\/\*{10})\s+){3}(?:\d+|\/\*{10}){1}\s*]{1}/g);

    if (!byteRangeStrings) {
        throw new SignPdfError(
            'No ByteRangeStrings found within PDF buffer',
            SignPdfError.TYPE_PARSE,
        );
    }

    const byteRangePlaceholder = byteRangeStrings.find((s) => s.includes(`/${DEFAULT_BYTE_RANGE_PLACEHOLDER}`));
    const byteRanges = byteRangeStrings.map((brs) => brs.match(/[^[\s]*(?:\d|\/\*{10})/g));

    return {
        byteRangePlaceholder,
        byteRangeStrings,
        byteRanges,
    };
};

export default findByteRange;
