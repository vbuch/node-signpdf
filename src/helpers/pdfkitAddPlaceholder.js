import {DEFAULT_BYTE_RANGE_PLACEHOLDER, DEFAULT_SIGNATURE_LENGTH, SUBFILTER_ADOBE_PKCS7_DETACHED} from './const';
// eslint-disable-next-line import/no-unresolved
import PDFKitReferenceMock from './pdfkitReferenceMock';
/**
 * Adds the objects that are needed for Adobe.PPKLite to read the signature.
 * Also includes a placeholder for the actual signature.
 * Returns an Object with all the added PDFReferences.
 * @param {PDFDocument} pdf
 * @param {string} reason
 * @returns {object}
 */
const pdfkitAddPlaceholder = ({
    pdf,
    pdfBuffer,
    reason,
    contactInfo = 'emailfromp1289@gmail.com',
    name = 'Name from p12',
    location = 'Location from p12',
    signatureLength = DEFAULT_SIGNATURE_LENGTH,
    byteRangePlaceholder = DEFAULT_BYTE_RANGE_PLACEHOLDER,
    subFilter = SUBFILTER_ADOBE_PKCS7_DETACHED,
}) => {
    /* eslint-disable no-underscore-dangle,no-param-reassign */
    // Generate the signature placeholder
    const signature = pdf.ref({
        Type: 'Sig',
        Filter: 'Adobe.PPKLite',
        SubFilter: subFilter,
        ByteRange: [
            0,
            byteRangePlaceholder,
            byteRangePlaceholder,
            byteRangePlaceholder,
        ],
        Contents: Buffer.from(String.fromCharCode(0).repeat(signatureLength)),
        Reason: new String(reason), // eslint-disable-line no-new-wrappers
        M: new Date(),
        ContactInfo: new String(contactInfo), // eslint-disable-line no-new-wrappers
        Name: new String(name), // eslint-disable-line no-new-wrappers
        Location: new String(location), // eslint-disable-line no-new-wrappers
    });

    // Check if pdf already contains acroform field
    const acroFormPosition = pdfBuffer.lastIndexOf('/Type /AcroForm');
    const isAcroFormExists = acroFormPosition !== -1;
    let fieldIds = [];
    let acroFormId;

    if (isAcroFormExists) {
        let acroFormStart = acroFormPosition;
        // 10 is the distance between "/Type /AcroForm" and AcroFrom ID
        const charsUntilIdEnd = 10;
        const acroFormIdEnd = acroFormPosition - charsUntilIdEnd;
        // Let's find AcroForm ID by trying to find the "\n" before the ID
        // 12 is a enough space to find the "\n"
        // (generally it's 2 or 3, but I'm giving a big space though)
        const maxAcroFormIdLength = 12;
        let foundAcroFormId = '';
        let index = charsUntilIdEnd + 1;
        for (index; index < charsUntilIdEnd + maxAcroFormIdLength; index += 1) {
            const acroFormIdString = pdfBuffer
                .slice(acroFormPosition - index, acroFormIdEnd).toString();

            if (acroFormIdString[0] === '\n') {
                break;
            }

            foundAcroFormId = acroFormIdString;
            acroFormStart = acroFormPosition - index;
        }

        const pdfSlice = pdfBuffer.slice(acroFormStart);
        const acroForm = pdfSlice.slice(0, pdfSlice.indexOf('endobj')).toString();
        acroFormId = parseInt(foundAcroFormId);

        const acroFormFields = acroForm.slice(acroForm.indexOf('/Fields [') + 9, acroForm.indexOf(']'));
        fieldIds = acroFormFields
            .split(' ')
            .filter((element, i) => i % 3 === 0)
            .map((fieldId) => new PDFKitReferenceMock(fieldId));
    }
    const signatureName = 'Signature';

    // Generate signature annotation widget
    const widget = pdf.ref({
        Type: 'Annot',
        Subtype: 'Widget',
        FT: 'Sig',
        Rect: [0, 0, 0, 0],
        V: signature,
        T: new String(signatureName + (fieldIds.length + 1)), // eslint-disable-line no-new-wrappers
        F: 4,
        P: pdf.page.dictionary, // eslint-disable-line no-underscore-dangle
    });
    pdf.page.dictionary.data.Annots = [widget];
    // Include the widget in a page
    let form;

    if (!isAcroFormExists) {
        // Create a form (with the widget) and link in the _root
        form = pdf.ref({
            Type: 'AcroForm',
            SigFlags: 3,
            Fields: [...fieldIds, widget],
        });
    } else {
        // Use existing acroform and extend the fields with newly created widgets
        form = pdf.ref({
            Type: 'AcroForm',
            SigFlags: 3,
            Fields: [...fieldIds, widget],
        }, acroFormId);
    }
    pdf._root.data.AcroForm = form;

    return {
        signature,
        form,
        widget,
    };
    /* eslint-enable no-underscore-dangle,no-param-reassign */
};

export default pdfkitAddPlaceholder;
