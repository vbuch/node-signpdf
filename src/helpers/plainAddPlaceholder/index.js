import PDFObject from '../pdfkit/pdfobject';
import PDFKitReferenceMock from '../pdfkitReferenceMock';
import removeTrailingNewLine from '../removeTrailingNewLine';
import {DEFAULT_SIGNATURE_LENGTH, SUBFILTER_ADOBE_PKCS7_DETACHED} from '../const';
import pdfkitAddPlaceholder from '../pdfkitAddPlaceholder';

import getIndexFromRef from './getIndexFromRef';
import readPdf from './readPdf';
import getPageRef from './getPageRef';
import createBufferRootWithAcroform from './createBufferRootWithAcroform';
import createBufferPageWithAnnotation from './createBufferPageWithAnnotation';
import createBufferTrailer from './createBufferTrailer';

const isContainBufferRootWithAcroform = (pdf) => {
    const bufferRootWithAcroformRefRegex = /\/AcroForm\s+(\d+\s\d+\sR)/g;
    const match = bufferRootWithAcroformRefRegex.exec(pdf.toString());

    return match != null && match[1] != null && match[1] !== '';
};

/**
 * Adds a signature placeholder to a PDF Buffer.
 *
 * This contrasts with the default pdfkit-based implementation.
 * Parsing is done using simple string operations.
 * Adding is done with `Buffer.concat`.
 * This allows node-signpdf to be used on any PDF and
 * not only on a freshly created through PDFKit one.
 */
const plainAddPlaceholder = ({
    pdfBuffer,
    reason,
    contactInfo = 'emailfromp1289@gmail.com',
    name = 'Name from p12',
    location = 'Location from p12',
    signatureLength = DEFAULT_SIGNATURE_LENGTH,
    subFilter = SUBFILTER_ADOBE_PKCS7_DETACHED,
}) => {
    let pdf = removeTrailingNewLine(pdfBuffer);
    const info = readPdf(pdf);
    const pageRef = getPageRef(pdf, info);
    const pageIndex = getIndexFromRef(info.xref, pageRef);
    const addedReferences = new Map();

    const pdfKitMock = {
        ref: (input, additionalIndex) => {
            info.xref.maxIndex += 1;

            const index = additionalIndex != null ? additionalIndex : info.xref.maxIndex;

            addedReferences.set(index, pdf.length + 1); // + 1 new line

            pdf = Buffer.concat([
                pdf,
                Buffer.from('\n'),
                Buffer.from(`${index} 0 obj\n`),
                Buffer.from(PDFObject.convert(input)),
                Buffer.from('\nendobj\n'),
            ]);
            return new PDFKitReferenceMock(info.xref.maxIndex);
        },
        page: {
            dictionary: new PDFKitReferenceMock(
                pageIndex,
                {
                    data: {
                        Annots: [],
                    },
                },
            ),
        },
        _root: {
            data: {},
        },
    };

    const {
        form,
        widget,
    } = pdfkitAddPlaceholder({
        pdf: pdfKitMock,
        pdfBuffer,
        reason,
        contactInfo,
        name,
        location,
        signatureLength,
        subFilter,
    });

    if (!isContainBufferRootWithAcroform(pdf)) {
        const rootIndex = getIndexFromRef(info.xref, info.rootRef);
        addedReferences.set(rootIndex, pdf.length + 1);
        pdf = Buffer.concat([
            pdf,
            Buffer.from('\n'),
            createBufferRootWithAcroform(pdf, info, form),
        ]);
    }
    addedReferences.set(pageIndex, pdf.length + 1);
    pdf = Buffer.concat([
        pdf,
        Buffer.from('\n'),
        createBufferPageWithAnnotation(pdf, info, pageRef, widget),
    ]);

    pdf = Buffer.concat([
        pdf,
        Buffer.from('\n'),
        createBufferTrailer(pdf, info, addedReferences),
    ]);

    return pdf;
};

export default plainAddPlaceholder;
