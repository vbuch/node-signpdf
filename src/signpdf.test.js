import PDFDocument from 'pdfkit';
import forge from 'node-forge';
import fs from 'fs';
import signer from './signpdf';
import {pdfkitAddPlaceholder, extractSignature, plainAddPlaceholder} from './helpers';
import SignPdfError from './SignPdfError';

/**
 * Creates a Buffer containing a PDF.
 * Returns a Promise that is resolved with the resulting Buffer of the PDFDocument.
 * @returns {Promise<Buffer>}
 */
const createPdf = (params) => new Promise((resolve) => {
    const requestParams = {
        placeholder: {},
        text: 'node-signpdf',
        addSignaturePlaceholder: true,
        pages: 1,
        layout: 'portrait',
        ...params,
    };

    const pdf = new PDFDocument({
        autoFirstPage: false,
        size: 'A4',
        layout: requestParams.layout,
        bufferPages: true,
    });
    pdf.info.CreationDate = '';

    if (requestParams.pages < 1) {
        requestParams.pages = 1;
    }

    // Add some content to the page(s)
    for (let i = 0; i < requestParams.pages; i += 1) {
        pdf
            .addPage()
            .fillColor('#333')
            .fontSize(25)
            .moveDown()
            .text(requestParams.text)
            .save();
    }

    // Collect the ouput PDF
    // and, when done, resolve with it stored in a Buffer
    const pdfChunks = [];
    pdf.on('data', (data) => {
        pdfChunks.push(data);
    });
    pdf.on('end', () => {
        resolve(Buffer.concat(pdfChunks));
    });

    if (requestParams.addSignaturePlaceholder) {
        // Externally (to PDFKit) add the signature placeholder.
        const refs = pdfkitAddPlaceholder({
            pdf,
            pdfBuffer: Buffer.from([pdf]),
            reason: 'I am the author',
            ...requestParams.placeholder,
        });
        // Externally end the streams of the created objects.
        // PDFKit doesn't know much about them, so it won't .end() them.
        Object.keys(refs).forEach((key) => refs[key].end());
    }

    // Also end the PDFDocument stream.
    // See pdf.on('end'... on how it is then converted to Buffer.
    pdf.end();
});

describe('Test signing', () => {
    it('expects PDF to be Buffer', () => {
        try {
            signer.sign('non-buffer', Buffer.from(''));
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_INPUT);
            expect(e.message).toMatchSnapshot();
        }
    });
    it('expects P12 certificate to be Buffer', () => {
        try {
            signer.sign(Buffer.from(''), 'non-buffer');
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_INPUT);
            expect(e.message).toMatchSnapshot();
        }
    });
    it('expects PDF to contain a ByteRange placeholder', () => {
        try {
            signer.sign(Buffer.from('No BR placeholder\n%%EOF'), Buffer.from(''));
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_PARSE);
            expect(e.message).toMatchSnapshot();
        }
    });
    it('expects a reasonably sized placeholder', async () => {
        try {
            const pdfBuffer = await createPdf({
                placeholder: {
                    signatureLength: 2,
                },
            });

            const p12Buffer = fs.readFileSync(
                `${__dirname}/../resources/certificate.p12`,
            );

            signer.sign(pdfBuffer, p12Buffer);
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_INPUT);
            expect(e.message).toMatchSnapshot();
        }
    });
    it('signs input PDF', async () => {
        let pdfBuffer = await createPdf();
        const p12Buffer = fs.readFileSync(
            `${__dirname}/../resources/certificate.p12`,
        );

        pdfBuffer = signer.sign(pdfBuffer, p12Buffer);
        expect(pdfBuffer instanceof Buffer).toBe(true);

        const {signature, signedData} = extractSignature(pdfBuffer);
        expect(typeof signature === 'string').toBe(true);
        expect(signedData instanceof Buffer).toBe(true);
    });
    it('signs a landscape PDF', async () => {
        let pdfBuffer = await createPdf({layout: 'landscape'});
        const p12Buffer = fs.readFileSync(
            `${__dirname}/../resources/certificate.p12`,
        );

        pdfBuffer = signer.sign(pdfBuffer, p12Buffer);
        expect(pdfBuffer instanceof Buffer).toBe(true);

        const {signature, signedData} = extractSignature(pdfBuffer);
        expect(typeof signature === 'string').toBe(true);
        expect(signedData instanceof Buffer).toBe(true);
    });
    it('signs detached', async () => {
        const p12Buffer = fs.readFileSync(
            `${__dirname}/../resources/certificate.p12`,
        );

        let pdfBuffer = await createPdf({text: 'Some text'});
        signer.sign(pdfBuffer, p12Buffer);
        const signature1 = signer.lastSignature;

        pdfBuffer = await createPdf({text: 'some other text '.repeat(30)});
        signer.sign(pdfBuffer, p12Buffer);
        const signature2 = signer.lastSignature;

        expect(signature1).not.toBe(signature2);
        expect(signature1).toHaveLength(signature2.length);
    });
    it('signs a ready pdf', async () => {
        const p12Buffer = fs.readFileSync(
            `${__dirname}/../resources/certificate.p12`,
        );
        let pdfBuffer = fs.readFileSync(`${__dirname}/../resources/w3dummy.pdf`);
        pdfBuffer = plainAddPlaceholder({
            pdfBuffer,
            reason: 'I have reviewed it.',
            signatureLength: 1612,
        });
        pdfBuffer = signer.sign(pdfBuffer, p12Buffer);

        const {signature, signedData} = extractSignature(pdfBuffer);
        expect(typeof signature === 'string').toBe(true);
        expect(signedData instanceof Buffer).toBe(true);
    });
    it('Signs the resource from issue 158', async () => {
        const p12Buffer = fs.readFileSync(
            `${__dirname}/../resources/certificate.p12`,
        );
        let pdfBuffer = fs.readFileSync(`${__dirname}/../resources/issue-158-test.pdf`);
        pdfBuffer = plainAddPlaceholder({
            pdfBuffer,
            reason: 'I have reviewed it.',
            signatureLength: 1612,
        });
        pdfBuffer = signer.sign(pdfBuffer, p12Buffer);
        const {signature, signedData} = extractSignature(pdfBuffer);
        expect(typeof signature === 'string').toBe(true);
        expect(signedData instanceof Buffer).toBe(true);
    });
    it('signs a ready pdf that does not have Annots', async () => {
        const p12Buffer = fs.readFileSync(
            `${__dirname}/../resources/certificate.p12`,
        );
        let pdfBuffer = fs.readFileSync(
            `${__dirname}/../resources/no-annotations.pdf`,
        );
        pdfBuffer = plainAddPlaceholder({
            pdfBuffer,
            reason: 'I have reviewed it.',
            signatureLength: 1612,
        });
        pdfBuffer = signer.sign(pdfBuffer, p12Buffer);

        const {signature, signedData} = extractSignature(pdfBuffer);
        expect(typeof signature === 'string').toBe(true);
        expect(signedData instanceof Buffer).toBe(true);
    });
    it('signs a ready pdf that does not have metadata', async () => {
        const p12Buffer = fs.readFileSync(
            `${__dirname}/../resources/certificate.p12`,
        );
        let pdfBuffer = fs.readFileSync(
            `${__dirname}/../resources/no-metadata.pdf`,
        );

        pdfBuffer = plainAddPlaceholder({
            pdfBuffer,
            reason: 'I have reviewed it.',
            signatureLength: 1612,
        });
        pdfBuffer = signer.sign(pdfBuffer, p12Buffer);

        const {signature, signedData} = extractSignature(pdfBuffer);
        expect(typeof signature === 'string').toBe(true);
        expect(signedData instanceof Buffer).toBe(true);
    });
    it('signs a ready pdf two times', async () => {
        const secondP12Buffer = fs.readFileSync(
            `${__dirname}/../resources/withpass.p12`,
        );
        let signedPdfBuffer = fs.readFileSync(
            `${__dirname}/../resources/signed-once.pdf`,
        );
        signedPdfBuffer = plainAddPlaceholder({
            pdfBuffer: signedPdfBuffer,
            reason: 'second',
            location: 'test location',
            signatureLength: 1592,
        });
        signedPdfBuffer = signer.sign(signedPdfBuffer, secondP12Buffer, {
            passphrase: 'node-signpdf',
        });
        const {signature, signedData} = extractSignature(signedPdfBuffer, 2);
        expect(typeof signature === 'string').toBe(true);
        expect(signedData instanceof Buffer).toBe(true);
    });
    it('signs big PDF twice producing big AcroForm ID on the first time', async () => {
        let pdfBuffer = await createPdf({
            pages: 100,
        });
        const p12Buffer = fs.readFileSync(
            `${__dirname}/../resources/certificate.p12`,
        );

        pdfBuffer = signer.sign(pdfBuffer, p12Buffer);
        expect(pdfBuffer instanceof Buffer).toBe(true);

        const {signature, signedData} = extractSignature(pdfBuffer);
        expect(typeof signature === 'string').toBe(true);
        expect(signedData instanceof Buffer).toBe(true);

        const secondP12Buffer = fs.readFileSync(
            `${__dirname}/../resources/withpass.p12`,
        );
        pdfBuffer = plainAddPlaceholder({
            pdfBuffer,
            reason: 'second',
            location: 'test location',
            signatureLength: 1592,
        });
        pdfBuffer = signer.sign(pdfBuffer, secondP12Buffer, {
            passphrase: 'node-signpdf',
        });
        expect(pdfBuffer instanceof Buffer).toBe(true);

        const {
            signature: secondSignature,
            signedData: secondSignatureData,
        } = extractSignature(pdfBuffer, 2);
        expect(typeof secondSignature === 'string').toBe(true);
        expect(secondSignatureData instanceof Buffer).toBe(true);
    });
    it('signs a ready pdf containing a link', async () => {
        const p12Buffer = fs.readFileSync(
            `${__dirname}/../resources/certificate.p12`,
        );
        let pdfBuffer = fs.readFileSync(
            `${__dirname}/../resources/including-a-link.pdf`,
        );
        pdfBuffer = plainAddPlaceholder({
            pdfBuffer,
            reason: 'I have reviewed it.',
            location: 'some city',
            signatureLength: 1612,
        });
        pdfBuffer = signer.sign(pdfBuffer, p12Buffer);

        const {signature, signedData} = extractSignature(pdfBuffer);
        expect(typeof signature === 'string').toBe(true);
        expect(signedData instanceof Buffer).toBe(true);
    });
    it('signs with ca, intermediate and multiple certificates bundle', async () => {
        let pdfBuffer = await createPdf();
        const p12Buffer = fs.readFileSync(`${__dirname}/../resources/bundle.p12`);

        pdfBuffer = signer.sign(pdfBuffer, p12Buffer);
        expect(pdfBuffer instanceof Buffer).toBe(true);

        const {signature, signedData} = extractSignature(pdfBuffer);
        expect(typeof signature === 'string').toBe(true);
        expect(signedData instanceof Buffer).toBe(true);
    });
    it('signs with passphrased certificate', async () => {
        let pdfBuffer = await createPdf();
        const p12Buffer = fs.readFileSync(`${__dirname}/../resources/withpass.p12`);

        pdfBuffer = signer.sign(pdfBuffer, p12Buffer, {
            passphrase: 'node-signpdf',
        });
        expect(pdfBuffer instanceof Buffer).toBe(true);

        const {signature, signedData} = extractSignature(pdfBuffer);
        expect(typeof signature === 'string').toBe(true);
        expect(signedData instanceof Buffer).toBe(true);
    });
    it('errors on wrong certificate passphrase', async () => {
        const pdfBuffer = await createPdf();
        const p12Buffer = fs.readFileSync(`${__dirname}/../resources/withpass.p12`);

        try {
            signer.sign(pdfBuffer, p12Buffer, {passphrase: 'Wrong passphrase'});
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof Error).toBe(true);
            expect(e.message).toMatchSnapshot();
        }
    });
    it('errors when no matching certificate is found in bags', async () => {
        const pdfBuffer = await createPdf();
        const p12Buffer = fs.readFileSync(`${__dirname}/../resources/bundle.p12`);

        // Monkey-patch pkcs12 to return no matching certificates although bundle.p12 is correct.
        const originalPkcs12FromAsn1 = forge.pkcs12.pkcs12FromAsn1;
        let p12Instance;
        forge.pkcs12.pkcs12FromAsn1 = (...params) => {
            // This instance will be used for all non-mocked code.
            p12Instance = originalPkcs12FromAsn1(...params);

            return {
                ...p12Instance,
                getBags: ({bagType}) => {
                    if (bagType === forge.pki.oids.certBag) {
                        // Only mock this case.
                        // Make sure there will be no matching certificate.
                        return {
                            [forge.pki.oids.certBag]: [],
                        };
                    }
                    return p12Instance.getBags({bagType});
                },
            };
        };

        try {
            signer.sign(pdfBuffer, p12Buffer);
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_INPUT);
            expect(e.message).toMatchSnapshot();
        } finally {
            forge.pkcs12.pkcs12FromAsn1 = originalPkcs12FromAsn1;
        }
    });
    it('error when final key inside trailer dictionary is /Root', async () => {
        let pdfBuffer = fs.readFileSync(
            `${__dirname}/../resources/w3dummy-different-trailer.pdf`,
        );
        pdfBuffer = plainAddPlaceholder({
            pdfBuffer,
            reason: 'I have reviewed it.',
            signatureLength: 1612,
        });
        const trailer = pdfBuffer
            .slice(pdfBuffer.lastIndexOf('trailer'))
            .toString();
        // the trailer should contain only one startxref
        expect(trailer.match(/startxref/g).length).toBe(1);
    });
    it('expects siging to fail because of no byteRangePlaceholder available to sign', async () => {
        try {
            const pdfBuffer = fs.readFileSync(`${__dirname}/../resources/signed.pdf`);
            const p12Buffer = fs.readFileSync(
                `${__dirname}/../resources/certificate.p12`,
            );

            signer.sign(pdfBuffer, p12Buffer);
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_PARSE);
        }
    });
});
