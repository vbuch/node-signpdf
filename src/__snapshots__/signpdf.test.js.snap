// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test signing errors on wrong certificate passphrase 1`] = `"PKCS#12 MAC could not be verified. Invalid password?"`;

exports[`Test signing errors when no matching certificate is found in bags 1`] = `"Failed to find a certificate that matches the private key."`;

exports[`Test signing expects P12 certificate to be Buffer 1`] = `"p12 certificate expected as Buffer."`;

exports[`Test signing expects PDF to be Buffer 1`] = `"PDF expected as Buffer."`;

exports[`Test signing expects PDF to contain a ByteRange placeholder 1`] = `"No ByteRangeStrings found within PDF buffer"`;

exports[`Test signing expects a reasonably sized placeholder 1`] = `"Signature exceeds placeholder length: 3224 > 4"`;
